ui/input/src/voiceMoveHandler.ts

```diff
@@ -4,11 +4,43 @@ import { Dests } from 'chessground/types';
 import { sanWriter, SanToUci } from 'chess';
 import * as util from './handlerUtil';
 
+const substitutionsMap = {
+  a: ['a8', '8'],
+  '3': ['3e'],
+  '8': ['a8', 'a'],
+  c: ['ce'],
+  d: ['de'],
+  g: ['ge'],
+};
+const closestLegalUci = (input: string, legalSans?: SanToUci): Uci | null => {
+  if (!legalSans) return null;
+  for (const [key, substitutions] of Object.entries(substitutionsMap)) {
+    for (const substitution of substitutions) {
+      const substituted = input.replace(key, substitution);
+      // check if after substitution, we have a legal SAN move
+      if (legalSans[substituted]) return legalSans[substituted];
+      // check if after substitution, we have a legal UCI move
+      if (substituted.match(util.fullUciRegex)) return substituted;
+    }
+  }
+  return null;
+};
+
 export function makeVoiceHandler(ctrl: MoveCtrl): MoveHandler {
   let legalSans: SanToUci | undefined;
   function submit(v: string) {
+    if (v.match(util.partialMoveRegex)) {
+      ctrl.voice.partialMove(v);
+      return;
+    }
+    if (v.match(util.cancelRegex)) {
+      ctrl.voice.partialMove('');
+      return;
+    }
+
```

Comment:
Don't be alarmed if this entire function is hacked apart in the near future.  It's what I was talking about when I mentioned "tangled mass of ifs".

Reply:
Totally, I've already said my good byes 😄  Gave some thought to a FSM approach, but couldn't quite figure it out. I just had a single transition with all of that logic in it 😅